File: sdk/python/src/roder_sdk/agent.py
from __future__ import annotations

import asyncio
import inspect
from collections.abc import Awaitable, Callable
from typing import Any, cast

from .client import RoderRpcClient
from .events import EventMode
from .run import RoderRun
from .transports import InMemoryTransport, LocalProcessTransport, RoderTransport, WebSocketTransport

ApprovalCallback = Callable[[Any], dict[str, Any] | Awaitable[dict[str, Any]]]


class RoderAgent:
    def __init__(
        self,
        transport: RoderTransport,
        *,
        cwd: str | None = None,
        model: dict[str, str] | None = None,
        thread_id: str | None = None,
        workspace_id: str | None = None,
        tool_allowlist: list[str] | None = None,
        instructions: str | None = None,
        runner: dict[str, Any] | None = None,
        approvals: dict[str, ApprovalCallback] | None = None,
        event_mode: EventMode = "permissive",
    ) -> None:
        self.transport = transport
        self.client = RoderRpcClient(transport)
        self.cwd = cwd
        self.model = model or {}
        self.thread_id = thread_id
        self.workspace_id = workspace_id
        self.tool_allowlist = tool_allowlist
        self.instructions = instructions
        self.runner = runner
        self.approvals = approvals or {}
        self.event_mode: EventMode = event_mode
        self._callback_task: asyncio.Task[None] | None = None

    @classmethod
    async def create(
        cls,
        *,
        local: dict[str, Any] | None = None,
        remote: dict[str, Any] | None = None,
        transport: RoderTransport | None = None,
        cwd: str | None = None,
        model: dict[str, str] | None = None,
        thread_id: str | None = None,
        workspace_id: str | None = None,
        tool_allowlist: list[str] | None = None,
        instructions: str | None = None,
        runner: dict[str, Any] | None = None,
        approvals: dict[str, ApprovalCallback] | None = None,
        event_mode: EventMode = "permissive",
    ) -> "RoderAgent":
        resolved = await _resolve_transport(local=local, remote=remote, transport=transport, cwd=cwd)
        agent = cls(
            resolved,
            cwd=cwd,
            model=model,
            thread_id=thread_id,
            workspace_id=workspace_id,
            tool_allowlist=tool_allowlist,
            instructions=instructions,
            runner=runner,
            approvals=approvals,
            event_mode=event_mode,
        )
        agent._start_callback_loop()
        return agent

    async def __aenter__(self) -> "RoderAgent":
        return self

    async def __aexit__(self, exc_type: object, exc: object, tb: object) -> None:
        await self.close()

    async def send(
        self,
        input: str | list[dict[str, Any]],
        *,
        developer_context: str | None = None,
    ) -> RoderRun:
        """
        developer_context is per-turn developer-authority context layered after
        the thread's developerInstructions for this turn only. Never persisted
        with the thread; resend it on each turn that needs it.
        """
        thread_id = self.thread_id or await self._start_thread()
        self.thread_id = thread_id
        params: dict[str, Any] = {"threadId": thread_id, "input": _normalize_input(input)}
        if developer_context is not None:
            params["developerContext"] = developer_context
        result = await self.client.call("turn/start", params)
        turn_id = _extract_id(result, "turn") or _extract_string(result, "turnId") or _extract_string(result, "id")
        if not turn_id:
            raise RuntimeError("turn/start response did not include a turn id")
        return RoderRun(self.client, thread_id, turn_id, event_mode=self.event_mode)

    async def list_models(self) -> Any:
        return await self.client.call("model/list")

    async def list_providers(self) -> Any:
        return await self.client.call("providers/list")

    async def read_thread(self, thread_id: str | None = None) -> Any:
        selected = thread_id or self.thread_id
        if not selected:
            raise RuntimeError("read_thread requires a thread id")
        return await self.client.call("thread/read", {"threadId": selected})

    async def list_threads(self) -> Any:
        return await self.client.call("thread/list")

    async def list_tools(self) -> Any:
        return await self.client.call("tools/list")

    async def list_commands(self) -> Any:
        return await self.client.call("commands/list")

    async def close(self) -> None:
        if self._callback_task:
            self._callback_task.cancel()
        await self.client.close()

    async def _start_thread(self) -> str:
        workspace_id = self.workspace_id or await self._resolve_workspace_id(self.cwd)
        params: dict[str, Any] = {
            "cwd": self.cwd,
            "model": self.model.get("id"),
            "modelProvider": self.model.get("provider"),
        }
        if self.tool_allowlist is not None:
            params["toolAllowlist"] = self.tool_allowlist
        if self.instructions is not None:
            params["developerInstructions"] = self.instructions
        if self.runner is not None:
            params["runner"] = self.runner
        params["workspaceId"] = workspace_id
        result = await self.client.call("thread/start", params)
        thread_id = _extract_id(result, "thread") or _extract_string(result, "threadId") or _extract_string(result, "id")
        if not thread_id:
            raise RuntimeError("thread/start response did not include a thread id")
        return thread_id

    async def _resolve_workspace_id(self, cwd: str | None) -> str:
        if not cwd:
            raise RuntimeError("starting a thread requires a workspace_id or a cwd to resolve one from")
        listed = await self.client.call("workspace/list", {})
        workspaces = listed.get("workspaces") if isinstance(listed, dict) else None
        for workspace in workspaces if isinstance(workspaces, list) else []:
            if not isinstance(workspace, dict):
                continue
            roots = workspace.get("roots")
            workspace_id = _extract_string(workspace, "id")
            if (
                workspace_id
                and isinstance(roots, list)
                and any(_extract_string(root, "path") == cwd for root in roots)
            ):
                return workspace_id
        created = await self.client.call("workspace/create", {"roots": [{"path": cwd}]})
        workspace_id = _extract_id(created, "workspace")
        if not workspace_id:
            raise RuntimeError("workspace/create response did not include a workspace id")
        return workspace_id

    def _start_callback_loop(self) -> None:
        if not self.approvals:
            return
        self._callback_task = asyncio.create_task(self._callback_loop())

    async def _callback_loop(self) -> None:
        async for notification in self.client.notifications():
            await self._handle_callback_notification(str(notification.get("method")), notification.get("params"))

    async def _handle_callback_notification(self, method: str, params: Any) -> None:
        if method == "thread/approvalRequested" and "on_tool_approval" in self.approvals:
            decision = await _maybe_await(self.approvals["on_tool_approval"](params))
            await self.client.call(
                "thread/resolve_approval",
                {
                    "approvalId": _extract_string(params, "approvalId"),
                    "approved": bool(decision.get("approved")),
                },
            )
        elif method == "thread/userInputRequested" and "on_user_input" in self.approvals:
            decision = await _maybe_await(self.approvals["on_user_input"](params))
            await self.client.call(
                "thread/resolve_user_input",
                {"requestId": _extract_string(params, "requestId"), "answers": decision.get("answers")},
            )
        elif method == "thread/planExitRequested" and "on_plan_exit" in self.approvals:
            decision = await _maybe_await(self.approvals["on_plan_exit"](params))
            await self.client.call(
                "thread/exit_plan",
                {
                    "requestId": _extract_string(params, "requestId"),
                    "approved": bool(decision.get("approved")),
                },
            )


async def _resolve_transport(
    *,
    local: dict[str, Any] | None,
    remote: dict[str, Any] | None,
    transport: RoderTransport | None,
    cwd: str | None,
) -> RoderTransport:
    if transport:
        return transport
    if remote:
        return await WebSocketTransport.connect(**remote)
    if local:
        return await LocalProcessTransport.create(
            command=local.get("command", "roder"),
            args=local.get("args"),
            cwd=local.get("cwd", cwd),
            env=local.get("env"),
        )
    return InMemoryTransport(
        lambda request: {
            "jsonrpc": "2.0",
            "id": request.get("id"),
            "error": {"code": -32000, "message": "no transport configured"},
        }
    )


def _normalize_input(input: str | list[dict[str, Any]]) -> list[dict[str, Any]]:
    return [{"type": "text", "text": input}] if isinstance(input, str) else input


def _extract_id(value: Any, key: str) -> str | None:
    if isinstance(value, dict) and isinstance(value.get(key), dict):
        return _extract_string(value[key], "id")
    return None


def _extract_string(value: Any, key: str) -> str | None:
    if isinstance(value, dict) and isinstance(value.get(key), str):
        return str(value[key])
    return None


async def _maybe_await(value: dict[str, Any] | Awaitable[dict[str, Any]]) -> dict[str, Any]:
    if inspect.isawaitable(value):
        return await cast(Awaitable[dict[str, Any]], value)
    return cast(dict[str, Any], value)

File: sdk/typescript/src/agent.ts
import { RoderRpcClient } from "./client.js";
import { RoderRun } from "./run.js";
import {
  InMemoryTransport,
  LocalProcessTransport,
  WebSocketTransport,
  type RoderTransport,
} from "./transports.js";
import type { EventMode } from "./events.js";

export interface RoderAgentOptions {
  local?: {
    cwd?: string;
    command?: string;
    args?: string[];
    env?: NodeJS.ProcessEnv;
    /** When false, the app-server receives exactly `env` instead of inheriting process.env. */
    inheritEnv?: boolean;
  };
  remote?: {
    url: string;
    token?: string;
    protocols?: string[];
  };
  transport?: RoderTransport;
  cwd?: string;
  model?: {
    provider?: string;
    id?: string;
    reasoning?: string;
  };
  /** Per-thread tool filter applied on top of the server's runtime allowlist. */
  toolAllowlist?: string[];
  /** Host instructions layered under the harness system prompt on every turn. */
  instructions?: string;
  /** Host-executed tools advertised to the model; calls arrive via onToolExecute. */
  externalTools?: RoderExternalTool[];
  /**
   * Binds the thread's native coding tools to a remote-runner workspace on the server. The config
   * is persisted with the thread, so secrets must reach the provider through its environment
   * (e.g. SAUNA_RUNNER_TOKEN), not this object.
   */
  runner?: RoderThreadRunner;
  /**
   * Executes a host tool call published by thread/toolExecutionRequested and replies via
   * tools/resolve. A thrown error resolves the call as an error result.
   */
  onToolExecute?(call: RoderExternalToolCall): Promise<ExternalToolResult> | ExternalToolResult;
  threadId?: string;
  workspaceId?: string;
  approvals?: RoderApprovals;
  eventMode?: EventMode;
}

export interface RoderExternalTool {
  name: string;
  description: string;
  /** JSON-schema for the tool arguments. */
  parameters: Record<string, unknown>;
}

export interface RoderThreadRunner {
  /** Installed remote-runner provider id (e.g. "sauna"). */
  providerId: string;
  /** Provider-specific destination config; persisted with the thread, so no secrets. */
  config?: Record<string, unknown>;
  /** Absolute path on the runner used as the thread's coding-tool workspace root. */
  workspace: string;
}

export interface RoderExternalToolCall {
  id: string;
  name: string;
  arguments: unknown;
}

export interface ExternalToolResult {
  output: string;
  isError?: boolean;
}

export interface RoderApprovals {
  onToolApproval?(request: unknown): Promise<ApprovalDecision> | ApprovalDecision;
  onUserInput?(request: unknown): Promise<UserInputDecision> | UserInputDecision;
  onPlanExit?(request: unknown): Promise<PlanExitDecision> | PlanExitDecision;
}

export interface ApprovalDecision {
  approved: boolean;
}

export interface UserInputDecision {
  answers: unknown;
}

export interface PlanExitDecision {
  approved: boolean;
}

export class RoderAgent {
  readonly client: RoderRpcClient;
  private threadId: string | undefined;
  private callbackLoopStarted = false;

  private constructor(
    private readonly transport: RoderTransport,
    private readonly options: RoderAgentOptions,
  ) {
    this.client = new RoderRpcClient(transport);
    this.threadId = options.threadId;
  }

  static async create(options: RoderAgentOptions): Promise<RoderAgent> {
    const transport = resolveTransport(options);
    const agent = new RoderAgent(transport, options);
    agent.startCallbackLoop();
    return agent;
  }

  async send(input: string | Array<Record<string, unknown>>, options: { eventMode?: EventMode } = {}): Promise<RoderRun> {
    const threadId = this.threadId ?? (await this.startThread());
    this.threadId = threadId;
    /**
     * Subscribe before turn/start: while the callback loop is active the hub
     * buffers nothing, so a turn/completed delivered in the same I/O chunk as
     * the turn/start response would be lost to a lazily subscribing run and
     * wait()/stream() would never terminate.
     */
    const notifications = this.client.notifications();
    try {
      const result = (await this.client.call("turn/start", {
        threadId,
        input: normalizeInput(input),
      })) as Record<string, unknown>;
      const turnId = extractId(result, "turn") ?? extractString(result, "turnId") ?? extractString(result, "id");
      if (!turnId) {
        throw new Error("turn/start response did not include a turn id");
      }
      return new RoderRun(this.client, threadId, turnId, {
        eventMode: options.eventMode ?? this.options.eventMode,
        notifications,
      });
    } catch (error) {
      void notifications[Symbol.asyncIterator]().return?.();
      throw error;
    }
  }

  async listModels(): Promise<unknown> {
    return this.client.call("model/list");
  }

  async listProviders(): Promise<unknown> {
    return this.client.call("providers/list");
  }

  async readThread(threadId = this.threadId): Promise<unknown> {
    if (!threadId) {
      throw new Error("readThread requires a thread id");
    }
    return this.client.call("thread/read", { threadId });
  }

  async listThreads(): Promise<unknown> {
    return this.client.call("thread/list");
  }

  async listTools(): Promise<unknown> {
    return this.client.call("tools/list");
  }

  async listCommands(): Promise<unknown> {
    return this.client.call("commands/list");
  }

  async close(): Promise<void> {
    await this.client.close();
  }

  private async startThread(): Promise<string> {
    const cwd = this.options.cwd ?? this.options.local?.cwd;
    const workspaceId = this.options.workspaceId ?? (await this.resolveWorkspaceId(cwd));
    const reasoning = this.options.model?.reasoning;
    const toolAllowlist = this.options.toolAllowlist;
    const instructions = this.options.instructions;
    const externalTools = this.options.externalTools;
    const runner = this.options.runner;
    const result = (await this.client.call("thread/start", {
      cwd,
      model: this.options.model?.id,
      modelProvider: this.options.model?.provider,
      ...(reasoning === undefined ? {} : { reasoning }),
      ...(toolAllowlist === undefined ? {} : { toolAllowlist }),
      ...(instructions === undefined ? {} : { developerInstructions: instructions }),
      ...(externalTools === undefined ? {} : { externalTools }),
      ...(runner === undefined ? {} : { runner }),
      workspaceId,
    })) as Record<string, unknown>;
    const threadId = extractId(result, "thread") ?? extractString(result, "threadId") ?? extractString(result, "id");
    if (!threadId) {
      throw new Error("thread/start response did not include a thread id");
    }
    return threadId;
  }

  private async resolveWorkspaceId(cwd: string | undefined): Promise<string> {
    if (!cwd) {
      throw new Error("starting a thread requires a workspaceId or a cwd to resolve one from");
    }
    const listed = (await this.client.call("workspace/list", {})) as Record<string, unknown>;
    const workspaces = Array.isArray(listed.workspaces) ? listed.workspaces : [];
    for (const workspace of workspaces) {
      if (!workspace || typeof workspace !== "object") {
        continue;
      }
      const roots = (workspace as Record<string, unknown>).roots;
      const id = extractString(workspace, "id");
      if (id && Array.isArray(roots) && roots.some((root) => extractString(root, "path") === cwd)) {
        return id;
      }
    }
    const created = (await this.client.call("workspace/create", {
      roots: [{ path: cwd }],
    })) as Record<string, unknown>;
    const workspaceId = extractId(created, "workspace");
    if (!workspaceId) {
      throw new Error("workspace/create response did not include a workspace id");
    }
    return workspaceId;
  }

  private startCallbackLoop(): void {
    if (this.callbackLoopStarted || (!this.options.approvals && !this.options.onToolExecute)) {
      return;
    }
    this.callbackLoopStarted = true;
    void (async () => {
      for await (const notification of this.transport.notifications()) {
        try {
          await this.handleCallbackNotification(notification.method, notification.params);
        } catch {
          /**
           * Resolution calls reject when the transport drops mid-callback or
           * the server refuses a stale request id; the server already times
           * the pending call out. Swallow so the loop keeps serving later
           * notifications instead of dying as an unhandled rejection that
           * crashes the host process.
           */
        }
      }
    })();
  }

  private async handleCallbackNotification(method: string, params: unknown): Promise<void> {
    /**
     * The server broadcasts every thread's notifications to every client on
     * the connection; only this agent's thread is ours to answer. Racing
     * another host's tools/resolve would silently feed it the wrong result
     * (first writer wins, the loser just sees resolved:false).
     */
    if (extractString(params, "threadId") !== this.threadId) {
      return;
    }
    if (method === "thread/toolExecutionRequested" && this.options.onToolExecute) {
      /**
       * Dispatched without awaiting: the server publishes a parallel tool
       * batch as back-to-back notifications and starts each call's
       * tools/resolve timeout at publication, so executing serially here
       * would burn call N+1's budget while call N runs. Each call resolves
       * independently; resolveExternalToolCall never rejects.
       */
      void this.resolveExternalToolCall(this.options.onToolExecute, params);
      return;
    }
    const approvals = this.options.approvals;
    if (!approvals) {
      return;
    }
    if (method === "thread/approvalRequested" && approvals.onToolApproval) {
      const decision = await approvals.onToolApproval(params);
      await this.client.call("thread/resolve_approval", {
        approvalId: extractString(params, "approvalId"),
        approved: decision.approved,
      });
    } else if (method === "thread/userInputRequested" && approvals.onUserInput) {
      const decision = await approvals.onUserInput(params);
      await this.client.call("thread/resolve_user_input", {
        requestId: extractString(params, "requestId"),
        answers: decision.answers,
      });
    } else if (method === "thread/planExitRequested" && approvals.onPlanExit) {
      const decision = await approvals.onPlanExit(params);
      await this.client.call("thread/exit_plan", {
        requestId: extractString(params, "requestId"),
        approved: decision.approved,
      });
    }
  }

  private async resolveExternalToolCall(
    onToolExecute: NonNullable<RoderAgentOptions["onToolExecute"]>,
    params: unknown,
  ): Promise<void> {
    const call = extractExternalToolCall(params);
    let result: ExternalToolResult;
    try {
      result = await onToolExecute(call);
    } catch (error) {
      result = { output: String(error), isError: true };
    }
    try {
      await this.client.call("tools/resolve", {
        requestId: extractString(params, "requestId"),
        output: result.output,
        isError: result.isError ?? false,
      });
    } catch {
      /**
       * The resolve rejects when the transport drops mid-call or the server
       * refuses a stale request id; the server already times the pending
       * call out. Swallow so the detached dispatch never surfaces as an
       * unhandled rejection that crashes the host process.
       */
    }
  }
}

function resolveTransport(options: RoderAgentOptions): RoderTransport {
  if (options.transport) {
    return options.transport;
  }
  if (options.remote) {
    return new WebSocketTransport(options.remote);
  }
  if (options.local) {
    return new LocalProcessTransport({
      command: options.local.command,
      args: options.local.args,
      cwd: options.local.cwd ?? options.cwd,
      env: options.local.env,
      inheritEnv: options.local.inheritEnv,
    });
  }
  return new InMemoryTransport((request) => ({
    jsonrpc: "2.0",
    id: request.id,
    error: { code: -32000, message: "no transport configured" },
  }));
}

function normalizeInput(input: string | Array<Record<string, unknown>>): Array<Record<string, unknown>> {
  return typeof input === "string" ? [{ type: "text", text: input }] : input;
}

function extractExternalToolCall(params: unknown): RoderExternalToolCall {
  const call =
    params && typeof params === "object" ? (params as Record<string, unknown>).call : undefined;
  return {
    id: extractString(call, "id") ?? "",
    name: extractString(call, "name") ?? "",
    arguments: call && typeof call === "object" ? (call as Record<string, unknown>).arguments : undefined,
  };
}

function extractId(value: Record<string, unknown>, key: string): string | undefined {
  const nested = value[key];
  return nested && typeof nested === "object" ? extractString(nested, "id") : undefined;
}

function extractString(value: unknown, key: string): string | undefined {
  if (!value || typeof value !== "object") {
    return undefined;
  }
  const candidate = (value as Record<string, unknown>)[key];
  return typeof candidate === "string" ? candidate : undefined;
}
